Reduce golint min_confidence to default 0.8

The current golint level of 1.0 causes several test cases to become errors, mostly due to capitalisation and docstrings.

However the default of 0.8 causes many more errors.

staticcheck complains about assert.go
```py
assert not (0 == 1)
```
becomes
```go
if !(!(0 == 1)) {
    panic("assert")
}
```
and staticcheck error:
> cases/assert.go:11:5: negating a boolean twice has no effect; is this a typo? (SA4013)

On macos, the default linters for golangci-lint are
```
Enabled by default linters:
deadcode: Finds unused code [fast: false, auto-fix: false]
errcheck: Errcheck is a program for checking for unchecked errors in go programs. These unchecked errors can be critical bugs in some cases [fast: false, auto-fix: false]
gosimple (megacheck): Linter for Go source code that specializes in simplifying a code [fast: false, auto-fix: false]
govet (vet, vetshadow): Vet examines Go source code and reports suspicious constructs, such as Printf calls whose arguments do not align with the format string [fast: false, auto-fix: false]
ineffassign: Detects when assignments to existing variables are not used [fast: true, auto-fix: false]
staticcheck (megacheck): Staticcheck is a go vet on steroids, applying a ton of static analysis checks [fast: false, auto-fix: false]
structcheck: Finds unused struct fields [fast: false, auto-fix: false]
typecheck: Like the front-end of a Go compiler, parses and type-checks Go code [fast: false, auto-fix: false]
unused (megacheck): Checks Go code for unused constants, variables, functions and types [fast: false, auto-fix: false]
varcheck: Finds unused global variables and constants [fast: false, auto-fix: false]
```

In addition to the error from staticcheck about assert.go, the other errors from `golangci-lint` are all from `deadcode` which is quite rightly pointing out that our test cases have dead code, c.f. https://github.com/adsharma/py2many/issues/123

coverage.go:
```
cases/coverage.go:17:6: `InlineEllipsis` is unused (deadcode)
func InlineEllipsis() {
     ^
cases/coverage.go:31:6: `InferBool` is unused (deadcode)
func InferBool(code int) bool {
     ^
```
global2.go:
```
cases/global2.go:9:5: `Code1` is unused (deadcode)
var Code1 int = 1
    ^
cases/global2.go:13:5: `LC` is unused (deadcode)
var LC = map[string]int{CodeB: Code0}
    ^
```

infer_ops.go:
```
cases/infer_ops.go:35:6: `Add1` is unused (deadcode)
func Add1(x int8, y int8) int16 {
     ^
cases/infer_ops.go:39:6: `Add2` is unused (deadcode)
func Add2(x int16, y int16) int32 {
     ^
cases/infer_ops.go:43:6: `Add3` is unused (deadcode)
func Add3(x int32, y int32) int64 {
     ^
cases/infer_ops.go:47:6: `Add4` is unused (deadcode)
func Add4(x int64, y int64) int64 {
     ^
cases/infer_ops.go:51:6: `Add5` is unused (deadcode)
func Add5(x uint8, y uint8) uint16 {
     ^
cases/infer_ops.go:55:6: `Add6` is unused (deadcode)
func Add6(x uint16, y uint16) uint32 {
     ^
cases/infer_ops.go:59:6: `Add7` is unused (deadcode)
func Add7(x uint32, y uint32) uint64 {
     ^
cases/infer_ops.go:63:6: `Add8` is unused (deadcode)
func Add8(x uint64, y uint64) uint64 {
     ^
cases/infer_ops.go:67:6: `Add9` is unused (deadcode)
func Add9(x int8, y uint16) uint32 {
     ^
cases/infer_ops.go:71:6: `Sub` is unused (deadcode)
func Sub(x int8, y int8) int8 {
     ^
cases/infer_ops.go:75:6: `Mul` is unused (deadcode)
func Mul(x int8, y int8) int16 {
     ^
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -41,8 +41,8 @@ jobs:
         run: |
           (go get github.com/hgfischer/go-iter \
            && go get github.com/google/go-cmp/cmp \
-           && go get golang.org/x/lint/golint \
            && go get github.com/adsharma/py2many/pygo/runtime \
+           && go get github.com/mgechev/revive \
           )
 
       - name: Install tox

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -30,6 +30,10 @@ jobs:
       - uses: julia-actions/setup-julia@v1
       - uses: fwilhe2/setup-kotlin@main
       - uses: jiro4989/setup-nim-action@v1
+      - uses: actions/setup-go@v2
+        # Needed as https://github.com/actions/setup-go/pull/89 is not merged
+        with:
+          go-version: '1.15'
       - uses: actions/setup-python@v2
         with:
           python-version: '3.8'

--- a/py2many/cli.py
+++ b/py2many/cli.py
@@ -58,6 +58,7 @@ from py2many.rewriters import (
 
 PY2MANY_DIR = pathlib.Path(__file__).parent
 ROOT_DIR = PY2MANY_DIR.parent
+CWD = pathlib.Path.cwd()
 
 
 def core_transformers(tree):
@@ -271,6 +272,12 @@ def dart_settings(args, env=os.environ):
 
 
 def go_settings(args, env=os.environ):
+    if os.path.exists(CWD / "revive.toml"):
+        revive_config = CWD / "revive.toml"
+    elif os.path.exists(PY2MANY_DIR / "revive.toml"):
+        revive_config = PY2MANY_DIR / "revive.toml"
+    else:
+        revive_config = None
     return LanguageSettings(
         GoTranspiler(),
         ".go",
@@ -279,7 +286,9 @@ def go_settings(args, env=os.environ):
         [GoNoneCompareRewriter(), GoVisibilityRewriter(), GoIfExpRewriter()],
         [infer_go_types],
         [GoMethodCallRewriter(), GoPropagateTypeAnnotation()],
-        linter=["golint", "-set_exit_status", "-min_confidence", "1.0"],
+        linter=(
+            ["revive", "--config", str(revive_config)] if revive_config else ["revive"]
+        ),
     )
 
 
@@ -296,7 +305,7 @@ def _get_all_settings(args, env=os.environ):
 
 
 def _relative_to_cwd(absolute_path):
-    return pathlib.Path(os.path.relpath(absolute_path, pathlib.Path.cwd()))
+    return pathlib.Path(os.path.relpath(absolute_path, CWD))
 
 
 def _process_once(settings, filename, outdir, env=None):
@@ -327,7 +336,8 @@ def _process_once(settings, filename, outdir, env=None):
         restore_cwd = False
         if settings.ext == ".kt" and output_path.parts[0] == "..":
             # ktlint can not handle relative paths starting with ..
-            restore_cwd = pathlib.Path.cwd()
+            restore_cwd = CWD
+
             os.chdir(output_path.parent)
             output_path = output_path.name
         cmd = _create_cmd(settings.formatter, filename=output_path)

--- a/revive.toml
+++ b/revive.toml
@@ -1,0 +1,30 @@
+ignoreGeneratedHeader = false
+severity = "warning"
+confidence = 0.8  # 0.1 also works
+errorCode = 1
+warningCode = 1
+
+# Enable default rules per https://github.com/mgechev/revive/blob/master/defaults.toml
+[rule.blank-imports]
+[rule.context-as-argument]
+[rule.context-keys-type]
+[rule.dot-imports]
+[rule.error-return]
+[rule.error-strings]
+[rule.error-naming]
+[rule.if-return]
+[rule.range]
+[rule.receiver-naming]
+[rule.time-naming]
+[rule.unexported-return]
+[rule.indent-error-flow]
+[rule.errorf]
+
+# Except these rules which are unlikely to be relevant
+# [rule.package-comments]
+# [rule.exported]
+
+# And these rules failing due to generator issues
+# [rule.var-naming] - fails due to __tmp..
+# [rule.var-declaration]
+# [rule.increment-decrement]

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -70,12 +70,14 @@ def in_cargo_toml(case: str):
     return f'path = "tests/expected/{case}.rs"' in CARGO_TOML
 
 
+def has_main_lines(lines):
+    return any("def main" in line or "__main__" in line for line in lines)
+
+
 def has_main(filename):
     with open(filename) as f:
         lines = f.readlines()
-    return bool(
-        [line in line for line in lines if "def main" in line or "__main__" in line]
-    )
+    return has_main_lines(lines)
 
 
 def get_exe_filename(case, ext):

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -11,6 +11,8 @@ from unittest_expander import foreach, expand
 
 from py2many.cli import _create_cmd, _get_all_settings, _relative_to_cwd, main
 
+import py2many.cli
+
 TESTS_DIR = Path(__file__).parent.absolute()
 ROOT_DIR = TESTS_DIR.parent
 BUILD_DIR = TESTS_DIR / "build"
@@ -57,7 +59,7 @@ EXTENSION_TEST_CASES = [
     if not item.stem.startswith("test_")
 ]
 
-EXPECTED_LINT_FAILURES = ["int_enum.go", "rect.go", "str_enum.go"]
+EXPECTED_LINT_FAILURES = []
 
 EXPECTED_COMPILE_FAILURES = []
 
@@ -100,6 +102,7 @@ class CodeGeneratorTests(unittest.TestCase):
     def setUp(self):
         os.makedirs(BUILD_DIR, exist_ok=True)
         os.chdir(BUILD_DIR)
+        py2many.cli.CWD = BUILD_DIR
 
     @foreach(sorted(LANGS))
     @foreach(sorted(TEST_CASES))

--- a/tests/test_unsupported.py
+++ b/tests/test_unsupported.py
@@ -23,6 +23,7 @@ from tests.test_cli import (
     KEEP_GENERATED,
     CXX,
     get_exe_filename,
+    has_main_lines,
 )
 
 ENV = {"rust": {"RUSTFLAGS": "--deny warnings"}}
@@ -176,9 +177,7 @@ TEST_ERROR_CASES = {
 
 def has_main(source):
     lines = source.splitlines()
-    return bool(
-        [line in line for line in lines if "def main" in line or "__main__" in line]
-    )
+    return has_main_lines(lines)
 
 
 @expand

--- a/tests/test_unsupported.py
+++ b/tests/test_unsupported.py
@@ -13,6 +13,8 @@ from unittest_expander import foreach, expand
 
 from py2many.cli import _get_all_settings, _relative_to_cwd, transpile
 
+import py2many.cli
+
 from tests.test_cli import (
     BUILD_DIR,
     GENERATED_DIR,
@@ -189,6 +191,7 @@ class CodeGeneratorTests(unittest.TestCase):
     def setUp(self):
         os.makedirs(BUILD_DIR, exist_ok=True)
         os.chdir(BUILD_DIR)
+        py2many.cli.CWD = BUILD_DIR
 
     @foreach(sorted(LANGS))
     @foreach(sorted(TEST_CASES.keys()))

